plutus-core/plutus-core/src/PlutusCore/Name/UniqueMap.hs

```diff
@@ -0,0 +1,115 @@
+module PlutusCore.Name.UniqueMap (
+  UniqueMap (..),
+insertByUnique,
+insertByName,
+singletonByName,
+insertNamed,
+insertByNameIndex,
+fromFoldable,
+fromUniques,
+fromNames,
+lookupUnique,
+lookupName,
+restrictKeys,
+foldr,
+lookupNameIndex,
+isEmpty,
+) where
+
+import Control.Lens (view)
+import Control.Lens.Getter ((^.))
+import Data.Coerce (Coercible, coerce)
+import Data.Foldable (foldl')
+import Data.IntMap.Strict qualified as IM
+import PlutusCore.Name.Unique (HasText (..), HasUnique (..), Named (Named), Unique (Unique))
+import PlutusCore.Name.UniqueSet (UniqueSet (UniqueSet))
+import Prelude hiding (foldr)
+
+-- | A mapping from uniques to values of type @a@.
+newtype UniqueMap unique a = UniqueMap
+  { unUniqueMap :: IM.IntMap a
+  }
+  deriving newtype (Show, Eq, Semigroup, Monoid, Functor)
+
+-- | Insert a value by a unique.
+insertByUnique ::
+  (Coercible unique Unique) =>
+  unique ->
+  a ->
+  UniqueMap unique a ->
+  UniqueMap unique a
+insertByUnique uniq = coerce . IM.insert (coerce uniq)
+
+-- | Insert a value by the unique of a name.
+insertByName :: (HasUnique name unique) => name -> a -> UniqueMap unique a -> UniqueMap unique a
+insertByName = insertByUnique . view unique
+
+singletonByName :: (HasUnique name unique) => name -> a -> UniqueMap unique a
+singletonByName n a = insertByName n a mempty
+
+-- | Insert a named value by the index of the unique of the name.
```

Comment:
I'm slightly unsure of your use of "index" here? I think you just mean by the unique? Applies throughout

Reply:
Oops, I forgot to fix the docs! Thanks for pointing it out.